From 0b63cdd02ab3cf259fdee65dd97247857126e8dc Mon Sep 17 00:00:00 2001
From: mmahesh <mmahesh@maxlinear.com>
Date: Tue, 15 Oct 2024 19:24:22 +0800
Subject: [PATCH] WLANRTSYS-85110 Assign invalid link id in the beginning

ISSUE Segfault seen while dereferencing drv from bss struct.

ROOTCAUSE Due to driver initalization failed in wpa_driver_nl80211_drv_init,
nl80211_remove_links is triggered in deinit flow.
As the Invalid link id is not yet set in init function and n_links remains zero,
nl80211_remove_links wrongly derefered bss->links[-1] and memset to zero.
This overwrites the value of drv pointer in bss struct.

FIX Set invalid link id in the beginning of wpa_driver_nl80211_drv_init.

Fixup to WLANRTSYS-60843 MLO changes: AP MLD configuration.
---
 src/drivers/driver_nl80211.c | 7 ++++---
 1 file changed, 4 insertions(+), 3 deletions(-)

diff --git a/src/drivers/driver_nl80211.c b/src/drivers/driver_nl80211.c
index 9f77ee8ae..99bea28ce 100644
--- a/src/drivers/driver_nl80211.c
+++ b/src/drivers/driver_nl80211.c
@@ -2457,6 +2457,10 @@ static void * wpa_driver_nl80211_drv_init(void *ctx, const char *ifname,
 	bss = drv->first_bss;
 	bss->drv = drv;
 	bss->ctx = ctx;
+#ifdef CONFIG_VENDOR_MXL
+	for (i = 0; i < MAX_NUM_MLD_LINKS; i++)
+		bss->links[i].link_id = NL80211_DRV_LINK_ID_NA;
+#endif /* CONFIG_VENDOR_MXL */
 
 	os_strlcpy(bss->ifname, ifname, sizeof(bss->ifname));
 	drv->monitor_ifidx = -1;
@@ -2514,9 +2518,6 @@ skip_wifi_status:
 	bss->n_links = 1;
 	os_memcpy(bss->flink->addr, bss->addr, ETH_ALEN);
 
-	for (i = 0; i < MAX_NUM_MLD_LINKS; i++)
-		bss->links[i].link_id = NL80211_DRV_LINK_ID_NA;
-
 	return bss;
 
 failed:
-- 
2.43.0

